46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+the the agent went wrong, the quote is not valid or one of the PCR based checks
+fail.
+
+This proposal changes that behavior and introduces the concept of severity
+levels of events and adds the option to provide more context for events.
+
+All parts that cause revocation events are tagged with an unique id and against
+those ids user scan specify a severity level and provide some context information.
+
+Instead of stop polling an agent if a failure occurred the agent is added back
+for checking if the failure is recoverable. If it is not a revocation event with
+the highest severity is generated.
+
+All the send events are visible in the verifier API for other systems to use.
+
+Keylime also gains the ability to specify context for a revocation event.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+The option to classify the revocation events opens the possibility that events
+are not handled. To mitigate this all revocation events that are not explicitly
+from the user classified are assigned the highest severity level.
+
+The same concept applies for revocation events that are caused failures that
+make the agent irrecoverable. Those also always need to be treated with the
+highest severity by default because otherwise some events wont be caught is a
+irrecoverable failure is triggered before.
+
+## Design Details
+We keep the current model of the states, but modify the behavior of the failure
+states. If we are in a failure state that is not irrecoverable the polling of
+the agent is stopped otherwise the agent is still added back for normal polling.
+
+A failure object is introduced. Which holds the all the revocation events that
+are being produced by the checks and if this any of the failures makes the agent
+not recoverable. Irrecoverable events currently are retry timeouts and quote
+validation.
+
+All parts of the validation process which is mainly (`check_qoute` and
+`check_pcrs`) will append all events to that object instead of returning false.
+If the validation process advance without a step that failed the failure object
+must be marked as irrecoverable and only then the function can return without
+validating further. The state of the agent after that should be
+`QUOTE_FAILED_IRRECOVERABLE`.
+
+The highest severity is 0 and the higher the number the number the lower the
```

Comment:
While I like the idea of being flexible, I'm not sure it's necessary.  I can't imagine real world scenarios with more than a handful of severity levels otherwise it will just be too operationally confusing. Also, I can see folks making mistakes in configuration because 0 is high and something like 10 is low. How about using something like the syslog severity text lables? It's well known and covers all kinds of scenarios:

- emerg
- alert
- crit
- err
- warning
- notice
- info
- debug

I could even see simplifying it more and merging emerg, alert, and crit into a single value like "crit"

Reply:
Yes, I can see that this might be too confusing and may lead to errors. Labels are more descriptive than numbers.

The main issue that I have with static labels is that you might use a label for a unintended purposes because you run out of labels. This will not be the case in most environments, but for us we want to use it for presorting events with a pretty fine granularity and I don't know if the proposed labels can represent accurately.

Another idea would be to use labels but make them configurable in the `keylime.conf`. Those labels would be strictly ordered by severity and are static during operation, but you still have the flexibility to add custom labels.

Reply:
I've changed the proposal to include user specified labels instead of using numbers.